src/recert.rs

```diff
@@ -44,18 +53,299 @@ pub(crate) async fn run(recert_config: &RecertConfig, cluster_crypto: &mut Clust
-            EtcdClient::connect([etcd_endpoint.as_str()], None)
-                .await
-                .context("connecting to etcd")?,
-        ),
-        None => None,
-    }));
+async fn is_encryption_enabled(in_memory_etcd_client: &Arc<InMemoryK8sEtcd>) -> Result<bool> {
+    let k8s_location = K8sResourceLocation::new(None, "APIServer", "cluster", "config.openshift.io/v1");
+
+    let cluster = get_etcd_json(in_memory_etcd_client, &k8s_location)
+        .await?
+        .context(format!("could not get {}", k8s_location))?;
+
+    if let Some(encryption_type) = cluster.pointer("/spec/encryption/type") {
+        let encryption_type = encryption_type.as_str().context("spec.encryption.type not a string")?;
+
+        Ok(encryption_type == "aescbc" || encryption_type == "aesgcm")
+    } else {
+        Ok(false)
+    }
+}
+
+async fn get_latest_encryption_config(recert_config: &RecertConfig) -> Result<String> {
+    let all_encryption_config_files = &recert_config
+        .cluster_customizations
+        .dirs
+        .iter()
+        .map(|dir| file_utils::globvec(dir, "**/encryption-config/encryption-config"))
+        .collect::<Result<Vec<_>>>()?
+        .into_iter()
+        .flatten()
+        .collect::<HashSet<_>>();
+
+    // Get latest revision from kube-apiserver-pod-(\d+) in the file path
+    let regex = regex::Regex::new(r"kube-apiserver-pod-(\d+)").context("compiling regex")?;
+    let captures = &all_encryption_config_files
+        .iter()
+        .filter_map(|file_pathbuf| {
+            let file_path = file_pathbuf.to_str()?;
+            Some((regex.captures(file_path)?[1].parse::<u32>().ok()?, file_pathbuf))
+        })
+        .collect::<HashSet<_>>();
+    let (_, latest_encryption_config) = captures
+        .iter()
+        .max_by_key(|(revision, _pathbuf)| revision)
+        .context("no kube-apiserver-pod-* found")?;
+
+    let latest_encryption_config_contents = &read_file_to_string(latest_encryption_config)
+        .await
+        .context("reading latest encryption-config")?;
+
+    Ok(latest_encryption_config_contents.to_string())
+}
+
+async fn get_apiserver_encryption_type(in_memory_etcd_client: &Arc<InMemoryK8sEtcd>) -> Result<String> {
+    let k8s_location = K8sResourceLocation::new(None, "APIServer", "cluster", "config.openshift.io/v1");
+
+    let apiserver = get_etcd_json(in_memory_etcd_client, &k8s_location)
+        .await?
+        .context(format!("could not get {}", k8s_location))?;
+
+    let encryption_type = apiserver
+        .pointer("/spec/encryption/type")
+        .context("no /spec/encryption/type")?
+        .as_str()
+        .context("spec.encryption.type not a string")?;
+
+    Ok(encryption_type.to_string())
+}
+
+async fn setup_etcd_client(recert_config: &RecertConfig) -> Result<(Arc<InMemoryK8sEtcd>, Option<EncryptionCustomizations>)> {
+    let mut in_memory_etcd_client = get_etcd_endpoint(recert_config, None, None).await?;
 
     log::info!("Connected to etcd");
```

Comment:
This log now happens before we even connect, so it's a bit misleadingingly successful